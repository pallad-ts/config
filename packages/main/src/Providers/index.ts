export * from './DefaultValueProvider';
export * from './EnvProvider';
export * from './FirstAvailableProvider';
export * from './FormattedStringProvider';
export * from './TransformProvider';
export * from './wrapWithDefaultAndTransformer';
