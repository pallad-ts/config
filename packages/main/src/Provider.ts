import { Either, fromTry, right } from "@sweet-monads/either";

import { protect, Secret, secret } from "@pallad/secret";
import { TypeCheck } from "@pallad/type-check";

import { ValueNotAvailable } from "./ValueNotAvailable";
import { runOnOptionalPromise } from "./common/runOnOptionalPromise";
import { OptionalPromise } from "./utils";

const CHECK = new TypeCheck("@pallad/config/Provider");

/**
 * Describes config value retrievable from other sources (env, env file, external storage etc)
 *
 * @public
 */
export abstract class Provider<TType> {
    constructor() {
        CHECK.assign(this);
    }

    /**
     * Returns value of provider
     */
    abstract getValue(): OptionalPromise<Provider.Value<TType>>;

    /**
     * Checks if given value is a Provider
     */
    static isType<T>(value: unknown): value is Provider<T> {
        return CHECK.isType(value);
    }

    transform<TNewType>(transformer: (value: TType) => TNewType): TransformProvider<TNewType, TType> {
        return new TransformProvider(this, transformer);
    }

    optional() {
        return this.defaultTo(undefined);
    }

    secret(): TransformProvider<Secret<TType>, TType> {
        return new TransformProvider(this, value => secret(value));
    }

    defaultTo<TNewType>(defaultValue: TNewType): DefaultValueProvider<TNewType, TType> {
        return new DefaultValueProvider(this, defaultValue);
    }
}

export namespace Provider {
    export type Value<T> = Either<Fail, T>;

    export type Fail = Fail.Entry | Fail.Entry[];
    export namespace Fail {
        export type Entry = ValueNotAvailable | Error;
    }
}

/**
 * Wrapper for other provider. Returns default value if given provider has no available value.
 *
 * @public
 */
export class DefaultValueProvider<T, TOriginal> extends Provider<T | TOriginal> {
    constructor(
        private provider: Provider<TOriginal>,
        private defaultValue: T
    ) {
        super();
    }

    getValue(): OptionalPromise<Provider.Value<T | TOriginal>> {
        return runOnOptionalPromise(this.provider.getValue(), result => {
            if (result.isLeft() && ValueNotAvailable.isType(result.value)) {
                return right<Provider.Fail, any>(this.defaultValue);
            }

            return result;
        });
    }
}

export class TransformProvider<TType, TSource> extends Provider<TType> {
    constructor(
        private provider: Provider<TSource>,
        private transformer: TransformProvider.Transformer<TType, TSource>
    ) {
        super();
    }

    getValue(): OptionalPromise<Provider.Value<TType>> {
        return runOnOptionalPromise(this.provider.getValue(), value => {
            return value.chain(value => {
                return fromTry(() => this.transformer(value));
            });
        });
    }
}

export namespace TransformProvider {
    export type Transformer<T, TSource = unknown> = (x: TSource, ...args: any[]) => T;
}
