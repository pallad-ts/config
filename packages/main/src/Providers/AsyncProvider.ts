import { Provider } from "../Provider";

export abstract class AsyncProvider<T> extends Provider<T> {
    private cache?: Promise<Provider.Value<T>>;

    getValue() {
        if (!this.cache) {
            this.cache = this.resolveValueAsync();
        }
        return this.cache;
    }

    abstract resolveValueAsync(): Promise<Provider.Value<T>>;
}
