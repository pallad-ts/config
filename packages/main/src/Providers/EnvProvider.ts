import {Provider} from "../Provider";

export class EnvProvider<T> extends Provider<string> {
    constructor(private key: string,
                private envs: typeof process['env'] = process.env) {
        super();
    }

    getDescription(): string {
        return 'ENV: ' + this.key;
    }

    isAvailable() {
        if (this.key in this.envs) {
            return this.envs[this.key] !== '';
        }
        return false;
    }

    protected retrieveValue() {
        return this.envs[this.key]!;
    }
}
