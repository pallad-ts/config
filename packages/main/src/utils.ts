import {Provider} from './Provider';

export type OptionalPromise<T> = T | Promise<T>;
export type UnwrapProvider<T> = T extends Provider<infer U> ? U : T;
