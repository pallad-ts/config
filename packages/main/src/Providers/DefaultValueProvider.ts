import {Provider} from "../Provider";
import {OptionalPromise} from '../utils';
import {isPromise} from '../common/isPromise';
import {runOnOptionalPromise} from '../common/runOnOptionalPromise';

/**
 * Wrapper for other provider. Returns default value if given provider has no available value.
 *
 * @public
 */
export class DefaultValueProvider<T, TOriginal> extends Provider<T | TOriginal> {
    constructor(private provider: Provider<TOriginal>, private defaultValue: T) {
        super();
    }

    getDescription(): string {
        return `${this.provider.getDescription()} - default: ${this.defaultValue}`;
    }

    isAvailable() {
        return true;
    }

    protected retrieveValue(): OptionalPromise<T | TOriginal> {
        return runOnOptionalPromise(
            this.provider.isAvailable(),
            isAvailable => {
                if (isAvailable) {
                    return this.provider.getValue() as any;
                }
                return this.defaultValue;
            }
        );
    }

    static optionalWrap<T, TOriginal>(provider: Provider<TOriginal>, defaultValue?: T) {
        if (defaultValue !== undefined) {
            return new DefaultValueProvider<T, TOriginal>(provider, defaultValue);
        }
        return provider;
    }
}
