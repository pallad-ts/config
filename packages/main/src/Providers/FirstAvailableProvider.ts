import {Provider} from "../Provider";
import {isPromise} from '../common/isPromise';
import {OptionalPromise} from '../utils';
import {runOnOptionalPromise} from '../common/runOnOptionalPromise';

export class FirstAvailableProvider<T extends Array<Provider<any>>> extends Provider<FirstAvailableProvider.Unwrap<T>> {
    private providers: Array<Provider<T>>;

    constructor(...providers: T) {
        super();
        this.providers = providers;
    }

    isAvailable() {
        return runOnOptionalPromise(this.getFirstAvailableProvider(), provider => !!provider);
    }

    private getFirstAvailableProvider(): OptionalPromise<Provider<any> | undefined> {
        const iterator = this.providers[Symbol.iterator]();
        for (const provider of iterator) {
            const isAvailable = provider.isAvailable();
            if (isPromise(isAvailable)) {
                return isAvailable.then(x => {
                    if (x) {
                        return provider;
                    }
                    return (async () => {
                        for (const provider of iterator) {
                            const isAvailable = await provider.isAvailable();
                            if (isAvailable) {
                                return provider;
                            }
                        }
                    })();
                });
            } else if (isAvailable) {
                return provider;
            }
        }
    }

    getDescription(): string {
        const desc = this.providers.map(x => x.getDescription()).join(', ');
        return `First available - ${desc}`;
    }

    retrieveValue() {
        return runOnOptionalPromise(
            this.getFirstAvailableProvider(),
            provider => {
                return provider!.getValue()
            }
        );
    }
}

export namespace FirstAvailableProvider {
    export type Unwrap<T> = T extends Array<Provider<infer U>> ? U : T;
}
