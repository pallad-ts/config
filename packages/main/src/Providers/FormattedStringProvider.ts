import {Provider} from "../Provider";
import {OptionalPromise} from '../utils';
import {isPromise} from '../common/isPromise';

function produce(strings: readonly string[], values: any[]) {
    let str = '';
    for (const [index, string] of strings.entries()) {
        str += `${string}${(values[index] || '')}`;
    }
    return str;
}

export class FormattedStringProvider extends Provider<string> {
    constructor(private strings: readonly string[], private values: any[]) {
        super();
    }

    getDescription(): string {
        const desc = produce(this.strings, this.values.map(x => Provider.is(x) ? `\${${x.getDescription()}}` : x));
        return `Formatted string ("${desc}")`;
    }

    isAvailable(): OptionalPromise<boolean> {
        const providersAvailability = this.getDependentProviders().map(x => x.isAvailable());
        const hasPromises = providersAvailability.some(isPromise);
        if (hasPromises) {
            return Promise.all(providersAvailability)
                .then(x => {
                    return x.every(x => x);
                })
        }
        return providersAvailability.every(x => x);
    }

    getDependentProviders(): Array<Provider<any>> {
        return this.values.filter(Provider.is);
    }

    retrieveValue(): OptionalPromise<string> {
        const values = this.values.map(x => Provider.is(x) ? x.getValue() : x)

        const hasPromises = values.some(isPromise);
        if (hasPromises) {
            return Promise.all(values)
                .then(x => {
                    return produce(this.strings, x);
                })
        }
        return produce(this.strings, values);
    }

    static create(strings: readonly string[], ...values: any[]) {
        if (values.some(Provider.is)) {
            return new FormattedStringProvider(strings, values);
        }
        return produce(strings, values);
    }
}
