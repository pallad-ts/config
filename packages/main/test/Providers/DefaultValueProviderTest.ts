import {DefaultValueProvider} from "@src/Providers";
import {DummyProvider} from "../dummies/DummyProvider";
import {Provider} from "@src/Provider";
import {assert, IsExact} from 'conditional-type-checks';
import {OptionalPromise} from '@src/utils';

describe('DefaultValueProvider', () => {
    const DEFAULT_VALUE = {default: 'value'} as const;

    const RAW_VALUE = {raw: 'value'} as const;

    function createProvider(provider: Provider<any>) {
        return new DefaultValueProvider(provider, DEFAULT_VALUE);
    }

    describe('getting value', () => {
        describe('from sync provider', () => {
            describe('where value is available', () => {
                const dep = createProvider(
                    new DummyProvider({isAvailable: true, value: RAW_VALUE, isAsync: false})
                );
                const value = dep.getValue();
                it('then returns provided provider value', () => {
                    expect(value)
                        .toStrictEqual(RAW_VALUE);
                });
            });

            describe('where value is not available', () => {
                const dep = createProvider(
                    new DummyProvider({isAvailable: false, value: RAW_VALUE, isAsync: false})
                );
                const value = dep.getValue();

                it('then returns default value', () => {
                    expect(value)
                        .toStrictEqual(DEFAULT_VALUE);
                });
            });
        });

        describe('from async provider', () => {
            describe('where value is available', () => {
                const dep = createProvider(
                    new DummyProvider({isAvailable: true, value: RAW_VALUE, isAsync: true})
                );
                const value = dep.getValue();
                it('then returns provided provider value', () => {
                    return expect(value)
                        .resolves
                        .toStrictEqual(RAW_VALUE);
                });
            });

            describe('where value is not available', () => {
                const dep = createProvider(
                    new DummyProvider({isAvailable: false, value: RAW_VALUE, isAsync: true})
                );
                const value = dep.getValue();

                it('then returns default value', () => {
                    return expect(value)
                        .resolves
                        .toStrictEqual(DEFAULT_VALUE);
                });
            });
        });
    });

    describe('checking availability', () => {
        describe('is always available', () => {
            it('for sync providers', () => {
                expect(
                    createProvider(
                        new DummyProvider({value: RAW_VALUE, isAsync: false, isAvailable: true}),
                    ).isAvailable()
                )
                    .toBe(true);

                expect(
                    createProvider(
                        new DummyProvider({value: RAW_VALUE, isAsync: false, isAvailable: false}),
                    ).isAvailable()
                )
                    .toBe(true);
            });

            it('for async providers', () => {
                expect(
                    createProvider(
                        new DummyProvider({value: RAW_VALUE, isAsync: true, isAvailable: true}),
                    ).isAvailable()
                )
                    .toBe(true);

                expect(
                    createProvider(
                        new DummyProvider({value: RAW_VALUE, isAsync: true, isAvailable: false}),
                    ).isAvailable()
                )
                    .toBe(true);
            });
        });
    });

    it('description', () => {
        expect(
            createProvider(
                new DummyProvider({value: RAW_VALUE, isAsync: false, isAvailable: true, description: 'Test'})
            )
                .getDescription()
        )
            .toMatchSnapshot();

        expect(
            createProvider(
                new DummyProvider({value: RAW_VALUE, isAsync: false, isAvailable: true, description: 'Random description'})
            )
                .getDescription()
        )
            .toMatchSnapshot();
    })

    describe('optionalWrap', () => {
        const dep = new DummyProvider({
            value: RAW_VALUE,
            isAsync: false,
            isAvailable: true
        });

        it('returns original provider if default value is undefined', () => {
            const provider = DefaultValueProvider.optionalWrap(dep, undefined);
            expect(provider)
                .toStrictEqual(dep);
        });

        it('wraps provided provider with TransformerProvider', () => {
            const provider = DefaultValueProvider.optionalWrap(dep, DEFAULT_VALUE);
            expect(provider)
                .toStrictEqual(new DefaultValueProvider(dep, DEFAULT_VALUE));
        });
    });

    it('types', () => {
        const provider = new DefaultValueProvider(
            DummyProvider.sync<'foo'>({value: 'foo', isAvailable: true}),
            'bar' as const
        );

        const value = provider.getValue();
        assert<IsExact<typeof value, OptionalPromise<'foo' | 'bar'>>>(true);
    });
});
