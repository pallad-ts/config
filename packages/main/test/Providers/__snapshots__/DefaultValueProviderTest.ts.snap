// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultValueProvider description 1`] = `"Test - default: [object Object]"`;

exports[`DefaultValueProvider description 2`] = `"Random description - default: [object Object]"`;
